pallets/migrations/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+	/// A human-readable name for this migration. Also used as storage key.
+	fn friendly_name(&self) -> &str;
+
+	/// Perform the required migration and return the weight consumed.
+	///
+	/// Currently there is no way to migrate across blocks, so this method must (1) perform its full
+	/// migration and (2) not produce a block that has gone over-weight. Not meeting these strict
+	/// constraints will lead to a bricked chain upon a runtime upgrade because the parachain will
+	/// not be able to produce a block that the relay chain will accept.
+	fn migrate(&self, available_weight: Weight) -> Weight;
+}
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_std::prelude::*;
+
+	/// Pallet for migrations
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The list of migrations that will be performed
+		type MigrationsList: Get<Vec<Box<dyn Migration>>>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RuntimeUpgradeStarted(),
+		RuntimeUpgradeCompleted(Weight),
+		MigrationStarted(Vec<u8>),
+		MigrationCompleted(Vec<u8>, Weight),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// on_runtime_upgrade is expected to be called exactly once after a runtime upgrade.
+		/// We use this as a chance to flag that we are now in upgrade-mode and begin our
+		/// migrations.
+		fn on_runtime_upgrade() -> Weight {
+			log::warn!("Performing on_runtime_upgrade");
+
+			let mut weight: Weight = 0u64.into();
+			// TODO: derive a suitable value here, which is probably something < max_block
+			let available_weight: Weight = T::BlockWeights::get().max_block;
+
+			// start by flagging that we are not fully upgraded
+			<FullyUpgraded<T>>::put(false);
+			weight += T::DbWeight::get().writes(1);
+			Self::deposit_event(Event::RuntimeUpgradeStarted());
+
+			weight += perform_runtime_upgrades::<T>(available_weight.saturating_sub(weight));
+
+			if !<FullyUpgraded<T>>::get() {
+				log::error!(
+					"migrations weren't completed in on_runtime_upgrade(), but we're not
+				configured for multi-block migrations; state is potentially inconsistent!"
+				);
+			}
+
+			weight
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn is_fully_upgraded)]
+	/// True if all required migrations have completed
+	type FullyUpgraded<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn migration_state)]
+	/// MigrationState tracks the progress of a migration.
+	/// Maps name (Vec<u8>) -> whether or not migration has been completed (bool)
+	type MigrationState<T: Config> = StorageMap<_, Twox64Concat, Vec<u8>, bool, OptionQuery>;
```

Comment:
This storage item never has any items inserted and mapped to false. So it is more like a set of completed migrations. For that reason it is not necessary to have both `bool` and `OptionQuery`. (because `Some(false)` will never be returned).

You could either make it a `ValueQuery` which will return the `Default` value (false) for items not stored. Or you could make it store `()` and treat `Some(())` to mean the migration ran and `None` to mean it didn't.

Reply:
607e0ee